File: java/08_055.java
import java.util.*;

/**
 * Trie node for storing words
 */
class TrieNode {
   Map<Character, TrieNode> children;
   boolean isEndOfWord;
   int lastSearched;  // Timestamp for history
   
   public TrieNode() {
      children = new HashMap<>();
      isEndOfWord = false;
      lastSearched = -1;
   }
}

/**
 * Search suggestion system with history and dictionary support
 */
public class SearchSuggestionSystem {
   private TrieNode dictionaryTrie;
   private TrieNode historyTrie;
   private int searchCount;  // Timestamp for recency
   
   /**
    * Initialize with dictionary words
    */
   public SearchSuggestionSystem(List<String> dictionary) {
      dictionaryTrie = new TrieNode();
      historyTrie = new TrieNode();
      searchCount = 0;
      
      // Add dictionary words
      if (dictionary != null) {
         for (String word : dictionary) {
            addToDictionary(word);
         }
      }
   }
   
   /**
    * Add a word to the dictionary
    */
   public void addToDictionary(String word) {
      TrieNode node = dictionaryTrie;
      for (char c : word.toCharArray()) {
         node.children.putIfAbsent(c, new TrieNode());
         node = node.children.get(c);
      }
      node.isEndOfWord = true;
   }
   
   /**
    * Add words to search history
    */
   public void addToHistory(List<String> words) {
      for (String word : words) {
         searchCount++;
         TrieNode node = historyTrie;
         for (char c : word.toCharArray()) {
            node.children.putIfAbsent(c, new TrieNode());
            node = node.children.get(c);
         }
         node.isEndOfWord = true;
         node.lastSearched = searchCount;  // Update timestamp
         
         // Ensure word is in dictionary too
         addToDictionary(word);
      }
   }
   
   /**
    * Suggest words based on prefix
    */
   public List<String> suggest(String prefix) {
      List<String> results = new ArrayList<>();
      
      // Get matching words from history
      List<Pair<String, Integer>> historyMatches = 
            new ArrayList<>();
      collectSuggestions(historyTrie, prefix, 
            new StringBuilder(), historyMatches, true);
      
      // Get matching words from dictionary
      List<String> dictMatches = new ArrayList<>();
      collectSuggestions(dictionaryTrie, prefix, 
            new StringBuilder(), dictMatches, false);
      
      // Track seen words to avoid duplicates
      Set<String> seen = new HashSet<>();
      
      // Sort history matches by recency (newest first)
      historyMatches.sort((a, b) -> 
            Integer.compare(b.getValue(), a.getValue()));
      for (Pair<String, Integer> pair : historyMatches) {
         String word = pair.getKey();
         if (!seen.contains(word)) {
            results.add(word);
            seen.add(word);
         }
      }
      
      // Add dictionary matches alphabetically
      Collections.sort(dictMatches);
      for (String word : dictMatches) {
         if (!seen.contains(word)) {
            results.add(word);
            seen.add(word);
         }
      }
      
      return results;
   }
   
   /**
    * Navigate to prefix node and collect suggestions
    */
   private void collectSuggestions(TrieNode root, String prefix, 
         StringBuilder currentWord, List<?> results, 
         boolean isHistory) {
      TrieNode node = root;
      
      // Navigate to prefix node
      for (char c : prefix.toCharArray()) {
         if (!node.children.containsKey(c)) {
            return;  // Prefix not found
         }
         node = node.children.get(c);
         currentWord.append(c);
      }
      
      // DFS to collect all words with this prefix
      dfsCollect(node, new StringBuilder(currentWord), 
            results, isHistory);
   }
   
   /**
    * DFS to collect words with given prefix
    */
   @SuppressWarnings("unchecked")
   private void dfsCollect(TrieNode node, StringBuilder currentWord, 
         List<?> results, boolean isHistory) {
      if (node.isEndOfWord) {
         if (isHistory) {
            ((List<Pair<String, Integer>>) results).add(
               new Pair<>(currentWord.toString(), 
                     node.lastSearched));
         } else {
            ((List<String>) results).add(currentWord.toString());
         }
      }
      
      // Process children in alphabetical order
      List<Character> sortedChars = 
         new ArrayList<>(node.children.keySet());
      Collections.sort(sortedChars);
      
      for (char c : sortedChars) {
         currentWord.append(c);
         dfsCollect(node.children.get(c), currentWord, 
               results, isHistory);
         // Backtrack
         currentWord.deleteCharAt(currentWord.length() - 1);
      }
   }
   
   /**
    * Helper class for key-value pairs
    */
   static class Pair<K, V> {
      private K key;
      private V value;
      
      public Pair(K key, V value) {
         this.key = key;
         this.value = value;
      }
      
      public K getKey() { return key; }
      public V getValue() { return value; }
   }
}

File: data structures and algorithms/cs/08_055.cs
using System;
using System.Collections.Generic;

public class TrieNode 
{
   public Dictionary<char, TrieNode> Children { get; set; }
   public bool IsEndOfWord { get; set; }
   public int LastSearched { get; set; }

   public TrieNode() 
   {
      Children = new Dictionary<char, TrieNode>();
      IsEndOfWord = false;
      LastSearched = -1;
   }
}

public class SearchSuggestionSystem 
{
   private TrieNode dictionaryTrie;
   private TrieNode historyTrie;
   private int searchCount;

   public SearchSuggestionSystem(List<string> dictionary) 
   {
      dictionaryTrie = new TrieNode();
      historyTrie = new TrieNode();
      searchCount = 0;

      if (dictionary != null) 
      {
         foreach (var word in dictionary) 
         {
            AddToDictionary(word);
         }
      }
   }

   public void AddToDictionary(string word) 
   {
      var node = dictionaryTrie;
      foreach (var c in word) 
      {
         if (!node.Children.ContainsKey(c)) 
         {
            node.Children[c] = new TrieNode();
         }
         node = node.Children[c];
      }
      node.IsEndOfWord = true;
   }

   public void AddToHistory(List<string> words) 
   {
      foreach (var word in words) 
      {
         searchCount++;
         var node = historyTrie;
         foreach (var c in word) 
         {
            if (!node.Children.ContainsKey(c)) 
            {
               node.Children[c] = new TrieNode();
            }
            node = node.Children[c];
         }
         node.IsEndOfWord = true;
         node.LastSearched = searchCount;
         AddToDictionary(word);
      }
   }

   public List<string> Suggest(string prefix) 
   {
      var results = new List<string>();
      var historyMatches = new List<Pair<string, int>>();
      
      CollectSuggestions(
         historyTrie, 
         prefix, 
         new StringBuilder(), 
         historyMatches, 
         true);

      var dictMatches = new List<string>();
      CollectSuggestions(
         dictionaryTrie, 
         prefix, 
         new StringBuilder(), 
         dictMatches, 
         false);

      var seen = new HashSet<string>();

      historyMatches.Sort((a, b) => 
         b.Value.CompareTo(a.Value));
      foreach (var pair in historyMatches) 
      {
         var word = pair.Key;
         if (!seen.Contains(word)) 
         {
            results.Add(word);
            seen.Add(word);
         }
      }

      dictMatches.Sort();
      foreach (var word in dictMatches) 
      {
         if (!seen.Contains(word)) 
         {
            results.Add(word);
            seen.Add(word);
         }
      }

      return results;
   }

   private void CollectSuggestions(
      TrieNode root, 
      string prefix, 
      StringBuilder currentWord, 
      List<object> results, 
      bool isHistory) 
   {
      var node = root;

      foreach (var c in prefix) 
      {
         if (!node.Children.ContainsKey(c)) 
         {
            return;
         }
         node = node.Children[c];
         currentWord.Append(c);
      }

      DfsCollect(
         node, 
         new StringBuilder(currentWord.ToString()), 
         results, 
         isHistory);
   }

   private void DfsCollect(
      TrieNode node, 
      StringBuilder currentWord, 
      List<object> results, 
      bool isHistory) 
   {
      if (node.IsEndOfWord) 
      {
         if (isHistory) 
         {
            results.Add(new Pair<string, int>(
               currentWord.ToString(), 
               node.LastSearched));
         } 
         else 
         {
            results.Add(currentWord.ToString());
         }
      }

      var sortedChars = new List<char>(node.Children.Keys);
      sortedChars.Sort();

      foreach (var c in sortedChars) 
      {
         currentWord.Append(c);
         DfsCollect(
            node.Children[c], 
            currentWord, 
            results, 
            isHistory);
         currentWord.Length--;
      }
   }

   public class Pair<K, V> 
   {
      public K Key { get; set; }
      public V Value { get; set; }

      public Pair(K key, V value) 
      {
         Key = key;
         Value = value;
      }
   }
}